### ISSUE TITLE
Apparent memory leak (seen on 3.2.3) when using TextStim

### ISSUE THREAD

**Issue by eneuro-bci** [psychopy/psychopy#2649]
I noticed a memory leak during an experiment. Inspecting the code carefully I realized that changing the text property of TextStim objects on a 60FPS rendering loop is possibly one of the trigger factors. To test this hypothesis I used this simple script which uses a single TextStim to show the elapsed time on a full-screen window:

```
def run_sample_experiment():

    from time import perf_counter
    from psychopy import core, visual

    # Start main clock (as reference of t=0)
    program_start_time = perf_counter()

    # Create window and text stimuli
    stimuli_window = \
        visual.Window([1920, 1080], fullscr=True, winType='pyglet', screen=1, monitor='testMonitor', units='norm',
                      gammaErrorPolicy="ignore", color=(0, 0, 0), colorSpace="rgb")

    stimuli_text = visual.TextStim(win=stimuli_window, text=" ", units="norm", pos=(0, 0), height=0.08,
                                   color=(255, 0, 0), colorSpace="rgb")
    stimuli_text.setAutoDraw(True)
    elapsed_time = perf_counter() - program_start_time

    while elapsed_time <= 300:
        elapsed_time = perf_counter() - program_start_time
        elapsed_time_str = "{:.2f}".format(elapsed_time)
        stimuli_text.text = elapsed_time_str
        stimuli_window.flip()

    # Close window and exit program
    core.wait(1)
    stimuli_window.close()
    return True


if __name__ == '__main__':
    run_sample_experiment()
```

I provide 3 screenshots of the task manager showing the memory consumption of the Python process running the experiment for a couple of minutes. A more complex experiment (with many visual stimulus) has this leak effect is increased, taking about 8GB of memory in less than 40 minutes.

I have noticed this behavior on 2 different machines, both running Windows 10 64-bit Pro with the May 2019 update. Both machines have an identical Python 3.7.4 environment (using the official installer, not Anaconda). The list  of packages and its version is on the attached txt file. 

I'm not sure if this leak is directly associated with Psychopy 3.2.3 or any of its dependencies (Numpy, Pygame, Pyglet, etc) is leaking on the specific versions I have on my environment. 

This is a serious problem for my goal, for now my approach is to revert back to different versions of Psychopy to see if the leak still occurs, but any advice is greatly appreciated. If it happens that this is really a Psychopy leak, I hope this info suffice to create a bug fix (otherwise let me know)

Thanks

--
screenshots
![minute 1](https://user-images.githubusercontent.com/52809490/66693161-93e72180-ecd8-11e9-8e1b-1887c0cc619b.png)
![minute 3](https://user-images.githubusercontent.com/52809490/66693162-947fb800-ecd8-11e9-9b44-39d644577214.png)
![minute 5](https://user-images.githubusercontent.com/52809490/66693160-93e72180-ecd8-11e9-9a9a-68c731c4e1f3.png)

--
environment
[python env.txt](https://github.com/psychopy/psychopy/files/3720263/python.env.txt)

**Comment by eneuro-bci** [comment]
As an update to this issue: 
The behavior is observed on a few Psychopy versions, so now I'm thinking it could be a problem with my current environment (if someone can please verify by running the code snippet above, that would be great). Also, by switching to the Pygame back-end when creating the window, the memory consumption increases at a much slower rate

**Comment by m-macaskill** [comment]
What version of `pyglet` do you have?

There was a long-standing memory leak there which affected PsychoPy scripts with frequent text updates:

https://bitbucket.org/pyglet/pyglet/issues/66/memory-leak-in-fonttext

but it was supposed to have been addressed with a bug fix in 2016:

https://bitbucket.org/pyglet/pyglet/commits/0692f89955c46eff088e953cdf78e291013336bc

**Comment by eneuro-bci** [comment]
I currently use pyglet 1.3.2. and pygame 1.9.6. I will test the script on a couple of updated pyglet versions, but also I did noticed the issue when using pygame as well

**Comment by eneuro-bci** [comment]
for any version of pyglet higher than 1.4 pygletbackedn.py crashes on line 101:
``defDisp = pyglet.window.get_platform().get_default_display()``

traceback:

> File "C:\Python37\lib\site-packages\psychopy\visual\window.py", line 386, in __init__
    self.backend = backends.getBackend(win=self, *args, **kwargs)
  File "C:\Python37\lib\site-packages\psychopy\visual\backends\__init__.py", line 32, in getBackend
    return Backend(win, *args, **kwargs)
  File "C:\Python37\lib\site-packages\psychopy\visual\backends\pygletbackend.py", line 101, in __init__
    defDisp = pyglet.window.get_platform().get_default_display()
AttributeError: module 'pyglet.window' has no attribute 'get_platform'
Exception ignored in: <function Window.__del__ at 0x000001B0160FD678>
Traceback (most recent call last):
  File "C:\Python37\lib\site-packages\psychopy\visual\window.py", line 493, in __del__
    self.close()
  File "C:\Python37\lib\site-packages\psychopy\visual\window.py", line 1277, in close
    self.backend.close()  # moved here, dereferencing the window prevents
AttributeError: 'NoneType' object has no attribute 'close'

I changed defDisp to:
``
display = pyglet.canvas.get_display()
``

then it crashes again with traceback:

> File "C:\Python37\lib\site-packages\psychopy\visual\text.py", line 209, in __init__
    self.setText(text, log=False)
  File "C:\Python37\lib\site-packages\psychopy\visual\text.py", line 354, in setText
    setAttribute(self, 'text', text, log)
  File "C:\Python37\lib\site-packages\psychopy\tools\attributetools.py", line 141, in setAttribute
    setattr(self, attrib, value)
  File "C:\Python37\lib\site-packages\psychopy\tools\attributetools.py", line 32, in __set__
    newValue = self.func(obj, value)
  File "C:\Python37\lib\site-packages\psychopy\visual\text.py", line 345, in text
    self._setTextShaders(text)
  File "C:\Python37\lib\site-packages\psychopy\visual\text.py", line 360, in _setTextShaders
    self._pygletTextObj = pyglet.font.Text(
AttributeError: module 'pyglet.font' has no attribute 'Text'

The solution to this crash was not obvious to me at the moment, any info on it much appreciated

**Comment by hoechenberger** [comment]
@eneuro-bci Pyglet 1.4 is currently not supported

**Comment by TEParsons** [comment]
The two errors described above in pyglet 1.4 are now handled and PsychoPy ships with pyglet 1.4.11 on Windows and 1.5.27 on Mac, so if this memory leak was fixed in pyglet 1.4 then we should be good. Please do reopen if this isn't the case though.
